Add new sample sets

smattis opened the pull request:
Adds rectangle (hyperrectangle), Lp-ball, and cartesian sample sets. Allows the type of sample set to be included when saving/loading.

lcgraham commented on bet/sample.py:
```diff
@@ -49,14 +50,16 @@ def save_sample_set(save_set, file_name, sample_set_name=None):
         if curr_attr is not None:
             mdat[sample_set_name+attrname] = curr_attr
+    mdat[sample_set_name + '_sample_set_type'] = save_set.__class__.__name__
+    mdat[sample_set_name + '_p_norm'] = save_set._p_norm
```
Remove this line and add `_p_norm` to `vector_names` on line 118.

lcgraham commented on bet/sample.py:
```diff
@@ -81,18 +84,20 @@ def load_sample_set(file_name, sample_set_name=None):
-        loaded_set = sample_set(np.squeeze(mdat[sample_set_name+"_dim"]))
+        loaded_set = eval(mdat[sample_set_name + '_sample_set_type'][0])(
+            np.squeeze(mdat[sample_set_name+"_dim"]))
+        loaded_set.set_p_norm(mdat[sample_set_name + '_p_norm'])
```
Unnecessary if line 118 is changed.

smattis replied:
should be fixed now

lcgraham replied:
Line 88 can be removed now.

lcgraham commented on bet/sample.py:
```diff
@@ -1123,18 +1140,205 @@ class sample_set(voronoi_sample_set):
+        :type mins: interable with components of length dim
+
+        """
+        if len(maxes) != len(mins):
```
Shouldn't we be setting the ``_domain`` in this method as well? I'm a little confused about what's going here and why ``np.inf`` is used. Is there a reason you did not do::

    self._domain = np.array((maxes, mins)) #might need a transpose
    self._right = np.array(maxes)
    self._left = np.array(mins)
    values = .5*(np.array(maxes)+np.array(mins))

lcgraham replied:
Disregard, I understand now.

lcgraham commented on bet/sample.py:
```diff
@@ -1123,18 +1140,205 @@ class sample_set(voronoi_sample_set):
+        :rtype: tuple
+        :returns: (dist, ptr)
+        """
+        num = self.check_num()
```
I think this could be much shorter and done without a for-loop by using ``self.update_bounds(num)``. Ah I think I understand the extra "value" out at infinity now... But I think the cleaner solution would be to set ``pt`` to an empty array and ``dist`` to ``None`` if ``x[i]``is not in the rectangle. Is the extra point at infinity so that the methods in ``calculateP`` will work?

smattis replied:
The extra cell is so query makes sense. It is also written so that there
can be multiple hyper rectangles. Sorry, won't be at a computer for awhile

On Fri, Jun 17, 2016 at 6:52 PM Lindley Graham <notifications@github.com>
wrote:

> In bet/sample.py
> <https://github.com/UT-CHG/BET/pull/220#discussion_r67589295>:
>
> > +        values[-1,:] = np.inf
> > +        self._right[-1,:] = np.inf
> > +        self._left[-1,:] = -np.inf
> > +        self._width = self._right - self._left
> > +        self.set_values(values)
> > +
> > +    def query(self, x, k=1):
> > +        """
> > +        Identify which value points x are associated with for discretization.
> > +        :param x: points for query
> > +        :type x: :class:`numpy.ndarray` of shape ``(*, dim)``
> > +        :param int k: number of nearest neighbors to return
> > +        :rtype: tuple
> > +        :returns: (dist, ptr)
> > +        """
> > +        num = self.check_num()
>
> I think this could be much shorter and done without a for-loop by using
> self.update_bounds(num). Ah I think I understand the extra "value" out at
> infinity now... But I think the cleaner solution would be to set pt to an
> empty array and dist to None if x[i]is not in the rectangle. Is the extra
> point at infinity so that the methods in calculateP will work?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/UT-CHG/BET/pull/220/files/fb336500fcbfa45214e0cef8a5909a3ef83f0162#r67589295>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA84Cau_3WJVpZ5Gv9WoxB1sf_SACKm5ks5qMzMkgaJpZM4I4wJy>
> .
>

lcgraham replied:
I understand now.

lcgraham commented on bet/sample.py:
```diff
@@ -1123,18 +1140,205 @@ class sample_set(voronoi_sample_set):
+        r"""
+        
+        Exactly calculates the Lebesgue volume fraction of the cells.
+        """
```
I think I'm unsure of what this sample set is representing. Is it a single hyperrectangle? If so shouldn't there just be a single value in ``_values`` and then shouldn't ``_volumes = np.ones((1,))``?

lcgraham replied:
Please disregard earlier comment, this makes sense based on the set definition.

lcgraham commented on bet/sample.py:
```diff
@@ -1123,18 +1140,205 @@ class sample_set(voronoi_sample_set):
+        :param centers: centers of balls
+        :type centers: interable of shape (num-1, dim)
+        :param radii: radii of balls
+        :type raii: iterable of length num-1
```
spelling

smattis replied:
Also, read the docstring for a description of the geometry
On Fri, Jun 17, 2016 at 6:58 PM Lindley Graham <notifications@github.com>
wrote:

> In bet/sample.py
> <https://github.com/UT-CHG/BET/pull/220#discussion_r67589595>:
>
> > +    discretization containing a number of balls.
> > +
> > +    A series of n balls :math:`A_i \subset \Lambda` with
> > +    :math:`A_i \cap A_j = \emptyset`
> > +    for :math:`i \neq j`. The last entry represents the remainder
> > +    :math:`\Lambda \setminus ( \cup_{i-1}^n A_i)`.
> > +
> > +    """
> > +    def setup(self, centers, radii):
> > +        """
> > +        Initialize.
> > +
> > +        :param centers: centers of balls
> > +        :type centers: interable of shape (num-1, dim)
> > +        :param radii: radii of balls
> > +        :type raii: iterable of length num-1
>
> spelling
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/UT-CHG/BET/pull/220/files/fb336500fcbfa45214e0cef8a5909a3ef83f0162#r67589595>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA84CX1P84LIr86k_QZ_dJQwdmquBI6Iks5qMzSNgaJpZM4I4wJy>
> .
>

lcgraham replied:
I understand now.

lcgraham commented:
Please disregard a bunch of my comments about the hyperrectangle class after fb36650. I didn't understand that it was the union of disjoint rectangles.

lcgraham commented on bet/sample.py:
```diff
@@ -1123,18 +1140,205 @@ class sample_set(voronoi_sample_set):
+        :type raii: iterable of length num-1
+        
+        """
+        #self.p_norm = p_norm
```
``p_norm`` should be an optional argument.

smattis replied:
`set_p_norm` is now a function within the `sample_set_base` class. That is where it should be set.

lcgraham commented on bet/sample.py:
```diff
@@ -1123,18 +1140,205 @@ class sample_set(voronoi_sample_set):
+        Identify which value points x are associated with for discretization.
+        :param x: points for query
+        :type x: :class:`numpy.ndarray` of shape ``(*, dim)``
+        :param int k: number of nearest neighbors to return
```
A space is required between lines 1263 and 1264 and 1266 and 1267 for sphinx to work correctly. Please check other new docstrings.

lcgraham commented on bet/sample.py:
```diff
@@ -1123,18 +1140,205 @@ class sample_set(voronoi_sample_set):
+        num = self.check_num()
+        self._volumes = np.zeros((num, ))
+        domain_vol = np.product(self._domain[:, 1] - self._domain[:, 0])
+        #self._volumes[0:-1] = ((2.0*scipy.special.gamma(1.0/float(self._p_norm) + 1.0)*self._width[0:-1])**self._dim)/scipy.special.gamma(float(self._dim)/float(self._p_norm) + 1.0)
```
remove comment

lcgraham commented on bet/sample.py:
```diff
@@ -1123,18 +1140,205 @@ class sample_set(voronoi_sample_set):
         '_emulated_ii_ptr_local', '_emulated_oo_ptr', '_emulated_oo_ptr_local']
     #: List of attribute names for attributes that are
-    #: :class:`sample.sample_set``
+    #: :class:`sample.sample_set_base``
```
remove extra `

lcgraham commented on bet/sample.py:
```diff
@@ -1123,18 +1140,205 @@ class sample_set(voronoi_sample_set):
+
+class cartesian_sample_set(rectangle_sample_set):
+    """
+    Defines a hyperrectangle discretization based on a Cartesian grid.
```
Be explicit in this docstring about the fact that the samples can define rectangles whose union is a subset of the domain rather than the entire domain. Note in the docstring that the coordinates of the grid are the corners of the Voronoi cells not the centers.

lcgraham commented:
Looks good, with only minor changes needed. I initially misunderstood the geometry/definition of these sets hence my earlier comments. I think you should add the method ``reset_bounds`` and ``reset_bounds_local`` for these new classes since these sample_sets define ``_left``, ``_right``, and ``_width`` differently and ``update_bounds`` as currently written will mess that up. I need to think more about this. Once solution might be to store ``_maxes``, ``_mins``, ``_rect_width`` and use those instead of ``_left``, ``_right``, and ``_width`` but that might not be the best solution.

Some things in ``basicSampling`` and ``adaptiveSampling`` might not work with these new sample_sets. The methods will need to be updated and type checks will need to be modified depending on how the methods get updated. I will think about this later.

lcgraham commented:
Okay, so since `basicSampling` and `adaptiveSampling` both check to see if the `input_obj` is a `sample_set` object and the proposed new classes are NOT `sample_set` objects (they inherit from `sample_set_base`) everything is fine. 

Later, if we do want `basicSampling` and `adaptiveSampling` to handle these new classes then changes will need to be made, possibly to all classes involved.

I've marked which line comments can be ignored based on my previous misunderstanding.

lcgraham commented:
I suggest the following changes to the docstring for `query` add the following to all `query` docstrings:

    .. seealso::

        :meth:`scipy.spatial.KDTree.query`

Add the following to new `query` docstrings (for the new classes):

    Only returns the neighbors for which :math:`x_i \in A_k`.

In the scipy implementation `ptr` is sorted by `dist` which you do not do (either add this functionality or note this difference, I would suggest adding). Also in both `rectangle_sample_set` and `ball_sample_set` you assume that the balls/rectangles are disjoint, however, there is no internal check to make sure that they are disjoint. If you add the sorting capability then the balls need not be disjoint as their intersection is determined by a local Voronoi partition. In this case we would need to update the docstring to include the specification of A_i and A_j as the union of the definition of the ball and the local Voronoi diagram. These could just be new classes which inherit from rectangle_sample_set and ball_sample_set with the assumption/check for disjointness removed and the sorting added. Also ``append_values`` and ``append_values_local`` need to be updated for these new classes.

smattis commented:
I'm travelling today and had a super busy weekend, so I probably won't be able to make the changes until tomorrow. Considering appending values, I was thinking of maybe not allowing it for now. Maybe just return a warning and do nothing. For now, I was not thinking these should really be used with adaptive sampling, so I wasn't thinking they would need to be appended. 

I will clarify things more in the docstring. I really wrote these with the idea of making `query` be super efficient, especially in high dimensions. That's why I wanted to force them to be disjoint now, because doing the local Voronoi discretization defeats the purpose.  

Also, for now here `dist` is either 0 or inf (in or out). I like this for these disjoint sets, because the `_values` are not what are defining the sets, it is the bounds of the cell. I will add a check for sets that intersect and raise an exception in that case. I could definitely see us adding fancier types of sample sets that inherit some things from these set in the future.

smattis commented:
I took the maxes/mins idea from scipy.spatial.Rectangle (http://docs.scipy.org/doc/scipy-0.17.0/reference/generated/scipy.spatial.Rectangle.html). We should probably use some stuff from their for fancier rectangle sample sets, but it adds unneccessary complexity for querying simple disjoint hyperrectangles.

lcgraham commented:
All of what you've said sounds great to me. I think not allowing appending
values for now sounds good (returning a warning and doing nothing). I
mainly wanted to discuss how you intended for these new classes to be used
and fit into the current use paradigms. Now I much better understand your
intentions. Thanks for the scipy link I definitely agree that we should
take advantage for fancier rectangle sets.

On Mon, Jun 20, 2016 at 4:19 PM Steven Mattis <notifications@github.com>
wrote:

> I took the maxes/mins idea from scipy.spatial.Rectangle (
> http://docs.scipy.org/doc/scipy-0.17.0/reference/generated/scipy.spatial.Rectangle.html).
> We should probably use some stuff from their for fancier rectangle sample
> sets, but it adds unneccessary complexity for querying simple disjoint
> hyperrectangles.
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/UT-CHG/BET/pull/220#issuecomment-227256958>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AC3o7VS4lEkQK4osVFv4AZUSQp4ni_N5ks5qNvXDgaJpZM4I4wJy>
> .
>
-- 

Lindley Graham, Ph.D. 2015

lcgraham commented:
Resolves #178

smattis commented:
@lcgraham I think this is ready now. I have added warnings and other minor touches. I think a `reset_bounds` function is unneccesary for the current use, as it is not used in adaptive sampling. You could easily just define a new sample set. It is actually a somewhat complicated problem in high dimensions to test if hyperrectangles intersect, so I just added a warning for now. This could be adjusted in the future, but I really don't think it's worth the time now, because we will usually just be using that type of discretization with one hypperrectangle.

lcgraham commented on bet/sample.py:
```diff
@@ -1298,7 +1478,7 @@ def query(self, x, k=1):
         pt = (num - 1) * np.ones((x.shape[0], k), dtype=np.int)
         for i in range(num - 1):
-            in_rec = np.less_equal(linalg.norm(x-self._values[i,:], self._p_norm, axis=1), self._width[i])
+            in_rec = np.less(linalg.norm(x-self._values[i,:], self._p_norm, axis=1), self._radii[i])
```
Why not less_equal?
